File: matlab/write_mls.m
problems = {'small', 'small2', 'medium', 'medium2', ...
            'large', 'large2', 'very', 'very2'};

for k = 1 : length(problems)
  problem = problems{k};
  fprintf('processing problem %s\n', problem);
  [A, b] = hb_to_msm(sprintf('%s.hb', problem));

  % scale problem
  n = size(A, 2);
  s = full(sqrt(sum(A.^2, 1)))';
  s(s == 0) = 1;
  S = spdiags(1./s, 0, n, n);
  A = A * S;

  % scaled minimum least-squares solution
  fprintf('  scaled mls...\n');
  x = mls(A, b);
  fid = fopen(sprintf('%s_scaled_mls.txt', problem), 'w');
  fprintf(fid, '%21.15e\n', x);
  fclose(fid);
end

File: julia/write_mls.jl
using LinearAlgebra, SparseArrays, DelimitedFiles
using Krylov, Quadmath, HarwellRutherfordBoeing

function scale_ls!(A)
  n, m = size(A)
  s = ones(m)
  for j = 1 : m
    i = A.colptr[j]
    k = A.colptr[j+1] - 1
    nj = (i ≤ k) ? norm(A.nzval[i:k]) : 0.0
    if nj > 0.0
      A.nzval[i:k] ./= nj
      s[j] = nj
    end
  end
  return s
end

for pb in ("small", "small2", "medium", "medium2", "large", "large2", "very", "very2")
  print("$pb: ")
  A = RutherfordBoeingData("../rb/$(pb).rb").data
  b = HarwellBoeingMatrix("../hb/$(pb).hb").rhs[:]
  s = scale_ls!(A)
  A = Float128.(A)
  b = Float128.(b)
  M = A' * A
  rhs = A' * b
  x, stats = minres_qlp(M, rhs, rtol=Float128(1e-20), atol=Float128(1e-25))
  println("✓")
  x = Float64.(x)
  open("../mls/krylov/$(pb)_scaled_mls.txt", "w") do io
    writedlm(io, x)
  end
end
